databuilder/databuilder/extractor/dashboard/apache_superset/__init__.py

```diff
@@ -0,0 +1,174 @@
+import abc
+from functools import reduce
+from typing import (
+    Any, Dict, Generator, Iterator, List, Tuple,
+)
+
+import requests
+from bs4 import BeautifulSoup
+from dateutil import parser
+from pyhocon import ConfigFactory, ConfigTree
+
+from databuilder.extractor.base_extractor import Extractor
+
+type_fields_mapping = List[Tuple[str, str, Any, Any]]
+
+
+class ApacheSupersetBaseExtractor(Extractor):
+    """
+    Base class to create extractors pulling any dashboard metadata from Apache Superset.
+    """
+    APACHE_SUPERSET_PROTOCOL = 'apache_superset_protocol'
+    APACHE_SUPERSET_HOST = 'apache_superset_host'
+    APACHE_SUPERSET_PORT = 'apache_superset_port'
+    APACHE_SUPERSET_USER = 'apache_superset_user'
+    APACHE_SUPERSET_PASSWORD = 'apache_superset_password'
+    APACHE_SUPERSET_PAGE_SIZE = 'apache_superset_page_size'
+    APACHE_SUPERSET_EXTRACT_PUBLISHED_ONLY = 'apache_superset_extract_published_only'
+
+    DASHBOARD_GROUP_NAME = 'dashboard_group_name'
+    DASHBOARD_GROUP_ID = 'dashboard_group_id'
+    DASHBOARD_GROUP_DESCRIPTION = 'dashboard_group_description'
+
+    PRODUCT = 'product'
+    CLUSTER = 'cluster'
+
+    DRIVER_TO_DATABASE_MAPPING = 'driver_to_database_mapping'
+
+    DEFAULT_DRIVER_TO_DATABASE_MAPPING = {
+        'postgresql': 'postgres',
+        'mysql+pymysql': 'mysql'
+    }
+
+    DATABASE_TO_CLUSTER_MAPPING = 'database_to_cluster_mapping'  # map superset dbs to preferred clusters
+
+    DEFAULT_CONFIG = ConfigFactory.from_dict({
+        APACHE_SUPERSET_PROTOCOL: 'http',
+        APACHE_SUPERSET_HOST: 'localhost',
+        APACHE_SUPERSET_PORT: '8088',
+        APACHE_SUPERSET_USER: 'admin',
+        APACHE_SUPERSET_PASSWORD: 'admin',
+        APACHE_SUPERSET_PAGE_SIZE: 20,
+        APACHE_SUPERSET_EXTRACT_PUBLISHED_ONLY: False,
+        PRODUCT: 'apache_superset',
+        DASHBOARD_GROUP_DESCRIPTION: '',
+        DRIVER_TO_DATABASE_MAPPING: DEFAULT_DRIVER_TO_DATABASE_MAPPING,
+        DATABASE_TO_CLUSTER_MAPPING: {}
+    })
+
+    def init(self, conf: ConfigTree) -> None:
+        self.session = requests.Session()
+        self.conf = conf.with_fallback(ApacheSupersetBaseExtractor.DEFAULT_CONFIG)
+        self._extract_iter = self._get_extract_iter()
+
+        self.authenticate()
+
+    def get_scope(self) -> str:
+        return 'extractor.apache_superset'
+
+    def extract(self) -> Any:
+        try:
+            result = next(self._extract_iter)
+
+            return result
+        except StopIteration:
+            return None
+
+    def authenticate(self) -> None:
+        # first request is to get csrf_token
+        login_form = self.session.get(self.build_full_url('login'))
+
+        # get Cross-Site Request Forgery protection token
+        soup = BeautifulSoup(login_form.text, 'html.parser')
```

Comment:
Is this the only way to authenticate? It looks like the superset docs have a [security section](https://superset.apache.org/docs/rest-api) for logging in. Does this not work for your use-case?

Reply:
+1, I wonder whether we could get the API token through config/hcvault instead of by parsing some input?

Reply:
good point, will refactor it to use `security/login`